integration-tests/actions/vrf/vrfv2/contract_steps.go

```diff
@@ -933,85 +725,78 @@ func WaitForRequestAndFulfillmentEvents(
+	}
+	err = env.EVMClient.WaitForEvents()
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s, err: %w", vrfcommon.ErrWaitTXsComplete, err)
+	}
 	l.Info().
-		Str("Coordinator", coordinator.Address()).
-		Str("Total Payment", randomWordsFulfilledEvent.Payment.String()).
-		Str("TX Hash", randomWordsFulfilledEvent.Raw.TxHash.String()).
-		Str("Request ID", randomWordsFulfilledEvent.RequestId.String()).
-		Bool("Success", randomWordsFulfilledEvent.Success).
-		Uint64("BlockNumber", randomWordsFulfilledEvent.Raw.BlockNumber).
-		Str("BlockHash", randomWordsFulfilledEvent.Raw.BlockHash.String()).
-		Msg("RandomWordsFulfilled Event (TX metadata)")
-}
-
-func LogRandomWordsForcedEvent(
-	l zerolog.Logger,
-	vrfOwner contracts.VRFOwner,
-	randomWordsForcedEvent *vrf_owner.VRFOwnerRandomWordsForced,
-) {
-	l.Debug().
-		Str("VRFOwner", vrfOwner.Address()).
-		Uint64("Sub ID", randomWordsForcedEvent.SubId).
-		Str("TX Hash", randomWordsForcedEvent.Raw.TxHash.String()).
-		Str("Request ID", randomWordsForcedEvent.RequestId.String()).
-		Str("Sender", randomWordsForcedEvent.Sender.String()).
-		Msg("RandomWordsForced Event (TX metadata)")
+		Str("Coordinator", *testConfig.VRFv2.ExistingEnvConfig.ExistingEnvConfig.CoordinatorAddress).
+		Int("Number of Subs to create", numberOfSubToCreate).
+		Msg("Creating and funding subscriptions, deploying and adding consumers to subs")
+	subIDs, err := CreateFundSubsAndAddConsumers(
+		env,
+		big.NewFloat(*testConfig.VRFv2.General.SubscriptionFundingAmountLink),
+		linkToken,
+		coordinator,
+		consumers,
+		numberOfSubToCreate,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("err: %w", err)
+	}
+	return consumers, subIDs, nil
 }
 
-func logRandRequest(
-	l zerolog.Logger,
-	consumer string,
-	coordinator string,
-	subID uint64,
-	minimumConfirmations uint16,
-	callbackGasLimit uint32,
-	numberOfWords uint32,
-	randomnessRequestCountPerRequest uint16,
-	randomnessRequestCountPerRequestDeviation uint16,
-) {
-	l.Info().
-		Str("Consumer", consumer).
-		Str("Coordinator", coordinator).
-		Uint64("SubID", subID).
-		Uint16("MinimumConfirmations", minimumConfirmations).
-		Uint32("CallbackGasLimit", callbackGasLimit).
-		Uint32("NumberOfWords", numberOfWords).
-		Uint16("RandomnessRequestCountPerRequest", randomnessRequestCountPerRequest).
-		Uint16("RandomnessRequestCountPerRequestDeviation", randomnessRequestCountPerRequestDeviation).
-		Msg("Requesting randomness")
+func CancelSubsAndReturnFunds(ctx context.Context, vrfContracts *vrfcommon.VRFContracts, eoaWalletAddress string, subIDs []uint64, l zerolog.Logger) {
+	for _, subID := range subIDs {
+		l.Info().
+			Uint64("Returning funds from SubID", subID).
+			Str("Returning funds to", eoaWalletAddress).
+			Msg("Canceling subscription and returning funds to subscription owner")
+		pendingRequestsExist, err := vrfContracts.CoordinatorV2.PendingRequestsExist(ctx, subID)
+		if err != nil {
+			l.Error().Err(err).Msg("Error checking if pending requests exist")
```

Comment:
This can also happen because of a wonky RPC - maybe try to sleep for a few seconds and retry several times before giving up, just a simple retry mechanism? Otherwise, the funds will be locked in this subscription forever. The same comment applies to line 794 when you are trying to cancel the existing subscription. If you know that it exists, don't give up on the first sign of trouble.

Reply:
lets do this in a separate [PR](https://smartcontract-it.atlassian.net/browse/VRF-953)?

Reply:
Yes, of course, you can merge this one first.